pkg/distributor/forwarding/forwarding.go

```diff
@@ -311,7 +349,36 @@ func (f *forwarder) splitByTargets(targetEndpoint string, tsSliceIn []mimirpb.Pr
 		}
 	}
 
-	return tsToIngest, tsByTargets
+	return tsToIngest, tsByTargets, err
+}
+
+// samplesNeedFiltering takes a slice of samples and a timestamp before which samples should be filtered out,
+// it returns true if some samples are older than the given timestamp or false otherwise.
+// It assumes that the samples are sorted by timestamp.
+func samplesNeedFiltering(samples []mimirpb.Sample, dontForwardBefore int64) bool {
+	if len(samples) == 0 {
+		return false
+	}
+
+	return samples[0].TimestampMs < dontForwardBefore
+}
+
+// filterSamplesBefore filters a given slice of samples to only contain samples that have timestamps newer or equal to
+// the given timestamp. It relies on the samples being sorted by timestamp.
```

Comment:
> It relies on the samples being sorted by timestamp.

Why can we rely on that? Push requests may not have samples sorted.

Reply:
Until we added OOO support the ingester required the samples within one series to be sorted as well afaik, otherwise we would have dropped some of them, this follows the same requirement.
However, you're right that now it is possible that we'd forward sample samples which are older than the cutoff at which we're supposed to start dropping them. I'd prefer to not have to sort the samples, so how about we reverse this logic and scale the slice of samples from the beginning to the end and as soon as we see one which is too old we drop the rest? That way if a slice of samples isn't sorted then we'd drop too much as opposed to too little.

Reply:
this flips the logic as suggested and also adds a unit test for it: 7e9915c2b

Reply:
I'm not sure I follow the logic for reversed iteration, but let's keep it as-is. No need to complicate this code more for would have been rejected as out-of-order samples before.